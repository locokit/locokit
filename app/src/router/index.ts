import { createRouter, createWebHistory } from 'vue-router'
import { routes, handleHotUpdate } from 'vue-router/auto-routes'

const router = createRouter({
  history: createWebHistory(import.meta.env.BASE_URL),
  routes,
})

// This will update routes at runtime without reloading the page
if (import.meta.hot) {
  handleHotUpdate(router)
}

export default router
