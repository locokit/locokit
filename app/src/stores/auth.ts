import { defineStore } from 'pinia'
import { SERVICES } from '@locokit/definitions'
import { ref } from 'vue'
import { useSDK } from '@/composables/useSDK'
import { type UserData } from '@locokit/sdk'

export const useStoreAuth = defineStore('auth', () => {
  const authState = ref({
    loading: false,
    isAuthenticated: false,
    error: null as Error | null,
    user: {} as UserData,
  })

  const { sdkClient } = useSDK()

  async function authenticate(data: { email: string; password: string }) {
    authState.value.loading = true
    authState.value.error = null
    try {
      const res = await sdkClient.authenticate({
        strategy: 'local',
        email: data.email,
        password: data.password,
      })
      // const token = useCookie('token') // useCookie new hook in nuxt 3
      // token.value = res.accessToken
      authState.value.user = res.user
      // if (result.user.rules) lckAbilities.update(result.user.rules)
      authState.value.isAuthenticated = true
    } catch (err) {
      console.error(err)
      authState.value.isAuthenticated = false
      authState.value.error = err as Error
    }
    authState.value.loading = false
  }

  async function reAuthenticate() {
    authState.value.loading = true
    authState.value.error = null
    try {
      // Todo: Not working, impossible to send a request
      const res = await sdkClient.reAuthenticate()
      // // const token = useCookie('token') // useCookie new hook in nuxt 3

      if (res.user) {
        authState.value.user = res.user
        // token.value = res.accessToken
        // if (result.user.rules) lckAbilities.update(result.user.rules)
        authState.value.isAuthenticated = true
      } else {
        // token.value = null
        authState.value.isAuthenticated = false
      }
    } catch (err) {
      console.error(err)
      authState.value.isAuthenticated = false
      authState.value.error = err as Error
    }
    authState.value.loading = false
  }

  async function sendLinkToResetPassword(data: { email: string }) {
    authState.value.loading = true
    authState.value.error = null
    try {
      await sdkClient.service(SERVICES.AUTH_AUTHENTICATION).create({
        action: 'sendResetPwd',
        value: data,
      })
    } catch (err) {
      console.error(err)
      authState.value.error = err as Error
    }
    authState.value.loading = false
  }

  async function resetPasswordLong(data: { token: string; password: string }) {
    authState.value.loading = true
    authState.value.error = null
    try {
      await sdkClient.service(SERVICES.AUTH_AUTHENTICATION).create({
        action: 'resetPwdLong',
        value: data,
      })
    } catch (err) {
      console.error(err)
      authState.value.error = err as Error
    }
    authState.value.loading = false
  }

  async function verifySignupAndSetPassword(data: { token: string; password: string }) {
    authState.value.loading = true
    authState.value.error = null
    try {
      await sdkClient.service(SERVICES.AUTH_AUTHENTICATION).create({
        action: 'verifySignupSetPasswordLong',
        value: data,
      })
    } catch (err) {
      console.error(err)
      authState.value.error = err as Error
    }
    authState.value.loading = false
  }

  async function signUp(data: { username: string; email: string }) {
    authState.value.loading = true
    authState.value.error = null
    try {
      await sdkClient.service(SERVICES.AUTH_SIGNUP).create(data)
    } catch (err) {
      console.error(err)
      authState.value.error = err as Error
    }
    authState.value.loading = false
  }

  async function logout() {
    authState.value.loading = true
    await sdkClient.logout()
    // // const token = useCookie('token') // useCookie new hook in nuxt 3
    authState.value.isAuthenticated = false
    // token.value = null // clear the token cookie
    authState.value.user = null
    authState.value.loading = false
  }

  async function sendEmailVerifySignup(userEmail: string) {
    authState.value.loading = true
    authState.value.error = null
    try {
      await sdkClient.service(SERVICES.AUTH_MANAGEMENT).create({
        action: 'resendVerifySignup',
        value: { email: userEmail },
      })
    } catch (err) {
      console.error(err)
      authState.value.error = err as Error
    }
    authState.value.loading = false
  }

  async function patchCurrentUser(id: string, data = {}) {
    authState.value.loading = true
    authState.value.error = null
    try {
      const res = await sdkClient.service(SERVICES.CORE_USER).patch(id, data)
      authState.value.user = res
      authState.value.loading = false
      return res
    } catch (err) {
      console.error(err)
      authState.value.error = err as Error
    }
    authState.value.loading = false
  }

  async function updateEmail(user: { email: string; password: string; newEmail: string }) {
    authState.value.loading = true
    authState.value.error = null
    try {
      const res = await sdkClient.service(SERVICES.AUTH_MANAGEMENT).create({
        action: 'identityChange',
        value: {
          user: { email: user.email },
          password: user.password,
          changes: { email: user.newEmail },
        },
      })
      authState.value.loading = false
      return res
    } catch (err) {
      console.error(err)
      authState.value.error = err as Error
    }
    authState.value.loading = false
  }

  async function confirmUpdateEmail(token: string) {
    authState.value.loading = true
    authState.value.error = null
    try {
      const res = await sdkClient.service(SERVICES.AUTH_MANAGEMENT).create({
        action: 'verifySignupLong',
        value: token,
      })
      authState.value.loading = false
      return res
    } catch (err) {
      console.error(err)
      authState.value.error = err as Error
    }
    authState.value.loading = false
  }

  async function updatePassword(user: { email: string; password: string; newPassword: string }) {
    authState.value.loading = true
    authState.value.error = null
    try {
      const res = await sdkClient.service(SERVICES.AUTH_MANAGEMENT).create({
        action: 'passwordChange',
        value: {
          user: { email: user.email },
          oldPassword: user.password,
          password: user.newPassword,
        },
      })
      authState.value.loading = false
      return res
    } catch (err) {
      console.error(err)
      authState.value.error = err as Error
    }
    authState.value.loading = false
  }

  return {
    authState,
    authenticate,
    reAuthenticate,
    sendLinkToResetPassword,
    resetPasswordLong,
    verifySignupAndSetPassword,
    signUp,
    logout,
    sendEmailVerifySignup,
    patchCurrentUser,
    updateEmail,
    confirmUpdateEmail,
    updatePassword,
  }
})
