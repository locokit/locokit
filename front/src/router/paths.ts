export const ROUTES_PATH = {
  HOME: '/',

  /**
   * User paths
   */
  LOSTPASSWORD: '/lost-password',
  VERIFYSIGNUP: '/verify-signup',
  RESETPASSWORD: '/reset-password',
  PROFILE: '/profile',
  UPDATEEMAIL: '/update-email',
  SIGNUP: '/signup',

  /**
   * Workspace paths
   */
  WORKSPACE: '/workspace',
  VISUALIZATION: '/visualization',
  VISUALIZATION_PAGE: '/page',
  VISUALIZATION_PAGE_DETAIL: '/detail',
  DATABASE: '/database',
  DATABASETABLE: '/table',
  DATABASESCHEMA: '/schema',
  PROCESS: '/process',
  SETTINGS: '/settings',
  FILES: '/files',
  ACLSET: '/aclset',
  GROUP: '/group', // used in workspace admin AND admin
  CMS: '/cms',
  CMS_PAGE: '/page',
  CMS_PAGE_DETAIL: '/detail',

  /**
   * Admin paths
   */
  ADMIN: '/admin',
  USER: '/user',
}

export const ROUTES_NAMES = {
  HOME: 'Home',

  /**
   * User paths
   */
  LOSTPASSWORD: 'LostPassword',
  VERIFYSIGNUP: 'VerifySignup',
  RESETPASSWORD: 'ResetPassword',
  PROFILE: 'Profile',
  UPDATEEMAIL: 'UpdateEmail',
  SIGNUP: 'SignUp',

  /**
   * Workspace paths
   */
  WORKSPACELIST: 'WorkspaceList',
  WORKSPACE: 'Workspace',
  WORKSPACE_ADMIN: {
    SELF: 'WorkspaceAdmin',
    DATABASE: 'WorkspaceDatabase',
    DATABASETABLE: 'WorkspaceDatabaseTable',
    DATABASESCHEMA: 'WorkspaceDatabaseSchema',
    PROCESS: 'WorkspaceProcess',
    PROCESS_ADD: 'WorkspaceProcessAdd',
    PROCESS_DETAIL: 'WorkspaceProcessDetail',
    ACL: 'WorkspaceACL',
    ACL_ADD: 'WorkspaceACLAdd',
    ACL_DETAIL: 'WorkspaceACLDetail',
    CMS: 'WorkspaceCMS',
    CMS_PAGE: 'WorkspaceCMSPage',
    CMS_PAGE_DETAIL: 'WorkspaceCMSPageDetail',
    SETTINGS: 'WorkspaceSettings',
    FILES: 'WorkspaceFiles',
    GROUP: 'WorkspaceGroup',
    GROUP_ADD: 'WorkspaceGroupAdd',
    GROUP_DETAIL: 'WorkspaceGroupDetail',
  },

  VISUALIZATION: 'WorkspaceVisualization',
  WORKSPACE_VISUALIZATION: {
    PAGE: 'WorkspaceVisualizationPage',
    PAGE_DETAIL: 'WorkspaceVisualizationPageDetail',
  },

  /**
   * Admin paths
   */
  ADMIN: {
    SELF: 'Administration',
    USER: 'AdminUser',
    GROUP: 'AdminGroup',
    GROUP_DETAIL: 'AdminGroupDetail', // used in workspace admin AND admin
    GROUP_ADD: 'AdminGroupAdd', // used in workspace admin AND admin
  },

  /**
   * Other paths
   */
  ERROR_404: '404',
}
