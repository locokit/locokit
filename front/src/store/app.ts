import { lckClient } from '@/services/lck-api'
import { LckSettings } from '@/services/lck-api/definitions'

class AppData {
  hasBurgerMenu = false
  allowSignUp = false
}
export const appState: AppData = {
  hasBurgerMenu: false,
  allowSignUp: false,
}

export async function loadApplicationSettings () {
  try {
    const applicationSettings: LckSettings = await lckClient.service('settings').find()
    appState.allowSignUp = applicationSettings.allow_signup
  } catch (e) {
    console.error(e)
  }
}
