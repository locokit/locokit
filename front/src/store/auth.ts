import lckAbilities from '@/services/lck-abilities'
import { lckClient } from '@/services/lck-api'
import { LckGroup } from '@/services/lck-api/definitions'
import {
  USER_PROFILE,
} from '@locokit/lck-glossary'
import { BaseState } from './state'

class User {
  id!: number;
  email!: string;
  name!: string;
  createdAt!: string;
  isVerified!: boolean
  profile!: USER_PROFILE
  groups?: LckGroup[]
}

class AuthData {
  isAuthenticated = false
  user: User | null = null
  currentGroupId: string | null = null
  expiredToken = false
}

export class AuthDTO {
  email = ''
  password = ''
}

export class AuthState extends BaseState<AuthData> {}

export const authState: AuthState = {
  loading: false,
  error: null,
  data: {
    isAuthenticated: false,
    user: null,
    currentGroupId: null,
    expiredToken: false,
  },
}

export async function retrieveUserGroupsAndWorkspacesAndDatabases (id: string) {
  authState.data.user = await lckClient.service('user').get(id, {
    query: {
      $eager: 'groups.[aclset.[workspace.[databases]]]',
    },
  })
}

export async function reAuthenticate () {
  authState.loading = true
  authState.data.expiredToken = false
  try {
    const result = await lckClient.reAuthenticate()
    lckAbilities.update(result.user.rules)
    authState.data.isAuthenticated = true
    await retrieveUserGroupsAndWorkspacesAndDatabases(result.user?.id)
  } catch (error) {
    authState.data.isAuthenticated = false
  }
  authState.loading = false
}

export async function authenticate (data: AuthDTO) {
  authState.loading = true
  authState.error = null
  authState.data.expiredToken = false
  try {
    const result = await lckClient.authenticate({
      strategy: 'local',
      email: data.email,
      password: data.password,
    })
    if (result.user.rules) lckAbilities.update(result.user.rules)
    authState.data.isAuthenticated = true
    await retrieveUserGroupsAndWorkspacesAndDatabases(result.user?.id)
  } catch (error) {
    authState.data.isAuthenticated = false
    authState.error = error as Error
  }
  authState.loading = false
}

export function logout () {
  authState.data = {
    isAuthenticated: false,
    user: null,
    currentGroupId: null,
    expiredToken: false,
  }
  lckAbilities.update([])
  return lckClient.logout()
}

export function updateUsername (username: string) {
  if (authState.data.user) {
    authState.data.user.name = username
  }
}
