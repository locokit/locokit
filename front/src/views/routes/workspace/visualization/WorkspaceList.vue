<template>
  <layout-with-header>
    <div class="generic-view-container p-mx-auto">
      <h1 class="p-my-4 lck-color-title">
        {{ $t('pages.workspace.title') }}
      </h1>

      <div class="p-grid">
        <div v-if="loading" class="p-col-12 p-md-4 p-lg-3 workspaces-item">
          <div class="workspaces-button">
            <div class="workspaces-detail">
              <p-skeleton
                class="workspaces-detail-title"
                width="10rem"
                height="2rem"
              ></p-skeleton>
            </div>
          </div>
        </div>

        <p v-if="!loading && workspaces.length === 0" class="p-col-12">
          {{ $t('pages.workspace.noWorkspace') }}
        </p>

        <div v-for="workspace in workspaces" :key="workspace.id" class="p-col-12 p-md-4 p-lg-3 workspaces-item">
          <div
            class="workspaces-button p-mr-2"
            :style="{
              backgroundColor: workspace.backgroundColor || 'inherit'
            }"
          >
            <router-link
              class="workspaces-detail"
              :to="`${ROUTES_PATH.WORKSPACE}/${workspace.id}`"
              :style="{
                color: workspace.color || 'inherit'
              }"
            >
              <p class="workspaces-detail-title">{{ workspace.text }}</p>
            </router-link>

            <router-link
              v-if="workspace.isManager"
              class="workspaces-admin"
              :to="`${ROUTES_PATH.WORKSPACE}/${workspace.id}${ROUTES_PATH.ADMIN}`"
              :style="{
                color: workspace.color || 'inherit',
              }"
            >
              <i class="bi bi-sliders"></i>
            </router-link>

            <i v-if="workspace.icon" class="workspaces-icon bi" :class="workspace.icon" />
          </div>
        </div>

        <div
          v-if="$can('create', 'workspace')"
          :class="'p-col-12 p-md-4 p-lg-3 workspaces-item' + ( workspaces.length === 0 ? ' p-mx-auto' : '')"
        >
          <button class="workspaces-new" @click="dialogVisible = true">
            <i class="bi bi-file-plus workspaces-new-icon"></i>
            <p class="p-button p-button-sm">{{ $t('pages.workspace.form.new') }}</p>
          </button>
        </div>
      </div>

      <p-dialog
        :header="$t('pages.workspace.form.create')"
        :modal="true"
        :visible="dialogVisible"
        @update:visible="dialogVisible = false"
      >

        <lck-workspace-form
          :submitting="submitting"
          @cancel="dialogVisible = false"
          @input="createWorkspace"
        />

      </p-dialog>
    </div>
  </layout-with-header>
</template>

<script lang="ts">
import Vue from 'vue'
import { ROUTES_PATH, ROUTES_NAMES } from '@/router/paths'
import { AuthState, authState } from '@/store/auth'
import { ColorScheme, COLOR_SCHEME } from '@/services/lck-utils/color'

import Skeleton from 'primevue/skeleton'

import WorkspaceForm from '@/components/visualize/WorkspaceForm/WorkspaceForm.vue'

import { lckServices } from '@/services/lck-api'
import Dialog from 'primevue/dialog/Dialog'
import { LckDatabase, LckWorkspace } from '@/services/lck-api/definitions'
import WithHeader from '@/layouts/WithHeader.vue'

const WORKSPACE_ROLE = {
  OWNER: 'OWNER',
  ADMIN: 'ADMIN',
  MEMBER: 'MEMBER',
}

export default {
  name: 'WorkspaceList',
  components: {
    'layout-with-header': Vue.extend(WithHeader),
    'lck-workspace-form': Vue.extend(WorkspaceForm),
    'p-dialog': Vue.extend(Dialog),
    'p-skeleton': Vue.extend(Skeleton),
  },
  data (): {
    loading: boolean;
    submitting: boolean;
    dialogVisible: boolean;
    ROUTES_PATH: typeof ROUTES_PATH;
    ROUTES_NAMES: typeof ROUTES_NAMES;
    authState: AuthState;
    colorScheme: ColorScheme[];
    WORKSPACE_ROLE: typeof WORKSPACE_ROLE;
    newWorkspace: Partial<LckWorkspace>;
    newWorkspaceColorScheme: ColorScheme | null;
    workspaces: {
      id: string;
      text: string;
      documentation?: string;
      color?: string | null;
      backgroundColor?: string | null;
      icon?: string | null;
      isManager: boolean;
    }[];
    } {
    return {
      loading: false,
      submitting: false,
      dialogVisible: false,
      ROUTES_PATH,
      ROUTES_NAMES,
      authState,
      WORKSPACE_ROLE,
      colorScheme: COLOR_SCHEME,
      newWorkspace: {
        text: '',
        documentation: '',
        settings: {},
      },
      newWorkspaceColorScheme: null,
      workspaces: [],
    }
  },
  methods: {
    transformDatabases (groupId: string, databases: LckDatabase[], schema = false) {
      return databases.map(({ text, id }) => ({
        id,
        label: text,
        to: `${ROUTES_PATH.WORKSPACE}/${groupId}${ROUTES_PATH.DATABASE}/${id}${schema ? ROUTES_PATH.DATABASESCHEMA : ''}`,
      }))
    },
    onColorSelect (event: { value: ColorScheme | null}) {
      this.newWorkspaceColorScheme = event.value
      if (event.value) {
        if (!this.newWorkspace.settings) {
          this.newWorkspace.settings = {}
        }
        this.newWorkspace.settings.color = event.value.color
        this.newWorkspace.settings.backgroundColor = event.value.backgroundColor
      }
    },
    async createWorkspace (newWorkspace: LckWorkspace) {
      this.submitting = true
      try {
        await lckServices.workspace.create(newWorkspace)
        this.$toast.add({
          severity: 'success',
          summary: this.$t('pages.workspace.form.createdSummary'),
          detail: this.$t('pages.workspace.form.createdDetail'),
          life: 5000,
        })
        this.dialogVisible = false
        this.fetchUserGroups()
      } catch (error) {
        this.$toast.add({
          severity: 'error',
          summary: this.$t('error.http.' + error.code),
          detail: this.$t('error.lck.' + error.data.code),
          life: 5000,
        })
      }
      this.submitting = false
    },
    async fetchUserGroups () {
      this.loading = true
      const userWorkspaces = await lckServices.workspace.find({
        query: {
          $modify: 'ofUser',
          $limit: -1,
        },
      }) as LckWorkspace[]
      this.workspaces = userWorkspaces.map((w: LckWorkspace) => {
        const currentWorkspace = {
          id: w.id,
          text: w.text,
          documentation: w.documentation,
          icon: w.settings?.icon,
          color: w.settings?.color,
          backgroundColor: w.settings?.backgroundColor,
          isManager: false,
        }
        // eslint-disable-next-line no-unused-expressions
        w.aclsets?.forEach(function (aclset) {
          if (aclset.manager) {
            currentWorkspace.isManager = true
          }
        })
        return currentWorkspace
      })
      this.loading = false
    },
  },
  mounted () {
    this.fetchUserGroups()
  },
  /**
   * This hook is executed each time the WorkspaceList is matched by the vue-router
   * So, sometimes, we don't want this hook execute
   * => we check before executing we are on the "good" component...
   */
  async beforeRouteEnter (to, from, next) {
    if (to.name !== ROUTES_NAMES.WORKSPACELIST) next()
    const userWorkspacesAvailable = authState?.data?.user?.groups
    if (
      !userWorkspacesAvailable?.some(({ aclset }) => aclset?.manager) &&
      userWorkspacesAvailable?.length === 1
    ) {
      // only one workspace, user is not a member of a group-aclset manager
      // we redirect user on the visualization route
      authState.data.currentGroupId = userWorkspacesAvailable[0].id
      next({
        name: ROUTES_NAMES.WORKSPACE,
        params: {
          workspaceId: userWorkspacesAvailable[0].aclset?.workspace_id as string,
          groupId: authState.data.currentGroupId,
        },
      })
    } else {
      next()
    }
  },
}
</script>
<style lang="scss" scoped>
.workspaces {
  &-item {
    padding: .75rem;
  }

  &-button {
    position: relative;
    overflow: hidden;
    display: flex;
    flex-direction: column;
    justify-content: center;
    padding: var(--spacing);
    min-height: 10rem;
    box-sizing: border-box;
    background: var(--color-white);
    align-content: center;
    text-align: center;
    border-radius: var(--border-radius);
    color: var(--text-color);
    box-shadow: 0 1px 3px 2px rgba(141, 27, 27, 0.04);
    transition: box-shadow .3s;
    font-weight: var(--font-weight-bold);
  }

  &-detail {
    position: relative;
    z-index: 2;
    text-decoration: none;
    cursor: pointer;

    &-title {
      margin: 0 auto;
      font-size: 1.6rem;
      word-wrap: break-word;
    }

    &:hover {
      text-decoration: underline;
    }
  }

  &-admin {
    position: absolute;
    z-index: 3;
    top: var(--spacing);
    right: var(--spacing);
    cursor: pointer;
    text-decoration: none;

    &:before {
        position: absolute;
        content: '';
        display: block;
        width: 1.25rem;
        height: 1.25rem;
        border-radius: 3rem;
        padding: 2rem;
        opacity: 0.1;
        bottom: -1.125rem;
        right: -3.25rem;
        transition: ease background-color .2s;
      }

    &:hover {
      &:before {
        background: currentColor;
      }
    }
  }

  &-icon {
    position: absolute;
    z-index: 1;
    left: -.75rem;
    bottom: -.75rem;
    font-size: 8rem;
    opacity: .1;
  }

  &-new {
    padding: var(--spacing);
    width: 100%;
    min-height: 10rem;
    width: 100%;
    border-radius: var(--border-radius);
    border: dashed 2px var(--secondary-color-light);
    cursor: pointer;
    background-color: transparent;
    color: var(--primary-color);

    &-icon {
      font-size: 2rem;
      display: block;
      margin-bottom: 1rem;
    }

    &:hover {
      border-color: var(--primary-color);
      background-color: var(--primary-color-lighten);
    }

    &:focus {
      outline: none;
      background-color: var(--primary-color-lighten);

      .p-button {
        background-color: var(--primary-color-dark);
      }
    }
  }
}
</style>
