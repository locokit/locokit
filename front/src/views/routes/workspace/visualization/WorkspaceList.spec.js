/* eslint-disable @typescript-eslint/camelcase */

import { shallowMount, createLocalVue } from '@vue/test-utils'
import { USER_PROFILE } from '@locokit/lck-glossary'
import VueRouter from 'vue-router'

import { ROUTES_PATH, ROUTES_NAMES } from '@/router/paths'
import { authState } from '@/store/auth'

import WorkspaceList from './WorkspaceList.vue'
import { abilitiesPlugin } from '@casl/vue'
import lckAbilitiesService from '@/services/lck-abilities'

import Vue from 'vue'

// Mock glossary
jest.mock('@locokit/lck-glossary', () => ({
  USER_PROFILE: {
    ADMIN: 'ADMIN',
    SUPERADMIN: 'SUPERADMIN',
    CREATOR: 'CREATOR',
    USER: 'USER',
  },
}))

// Mock variables
const mockWorkspaceContent = {
  id: '0',
  text: 'workspace-title',
  createdAt: '2021-01-01T00:00:00.109Z',
  updatedAt: '2021-01-01T00:00:00.109Z',
  chapters: [
    {
      id: '1',
      text: 'Chapter 1',
      createdAt: '2020-11-02T16:11:03.109Z',
      updatedAt: '2020-12-23T10:56:11.789Z',
      workspace_id: '0',
      pages: [
        {
          id: '11',
          text: 'Page 1',
          createdAt: '2020-11-02T16:11:03.109Z',
          updatedAt: '2020-11-02T16:11:03.109Z',
          chapter_id: '1',
          position: 0,
          hidden: false,
        },
        {
          id: '12',
          text: 'Page 2',
          createdAt: '2020-11-02T16:11:03.109Z',
          updatedAt: '2020-11-02T16:11:03.109Z',
          chapter_id: '1',
          position: 1,
          hidden: true,
        },
      ],
    },
    {
      id: '2',
      text: 'Chapter 2',
      createdAt: '2020-11-02T16:11:03.109Z',
      updatedAt: '2020-12-23T10:56:11.789Z',
      workspace_id: '0',
    },
  ],
}

function mockDeepCloneObject (object) {
  return JSON.parse(JSON.stringify(object))
}

// Mock lck functions

jest.mock('@/services/lck-api', () => ({
  lckServices: {
    chapter: {
      patch: jest.fn((id, data) =>
        ({ ...mockWorkspaceContent.chapters.find(c => c.id === id), ...data })),
      create: jest.fn(data =>
        ({ ...mockWorkspaceContent.chapters[0], ...data, id: '2' })),
      remove: jest.fn(),
    },
    page: {
      patch: jest.fn((id, data) =>
        ({ ...mockWorkspaceContent.chapters[0].pages[0], ...data, id })),
      create: jest.fn(data =>
        ({ ...mockWorkspaceContent.chapters[0].pages[0], ...data, id: '13' })),
      remove: jest.fn(),
    },
    group: {
      find: jest.fn(() => []),
    },
    workspace: {
      find: jest.fn(() => [mockWorkspaceContent]),
    },
  },
  lckHelpers: {
    retrieveWorkspaceWithChaptersAndPages: () => mockDeepCloneObject(mockWorkspaceContent),
  },
}))

jest.mock('@/store/auth', () => ({
  authState: {
    data: {
      isAuthenticated: true,
      user: {
        profile: '',
        groups: [],
        rules: [],
      },
    },
  },
}))

jest.mock('@/router/paths', () => ({
  ROUTES_PATH: {
    WORKSPACE: '/workspace',
    VISUALIZATION: '/visualization',
  },
  ROUTES_NAMES: {
    VISUALIZATION: 'WorkspaceVisualization',
    PAGE: 'Page',
  },
}))

// Mock Vue components
jest.mock('@/views/routes/workspace/visualization/Page.vue', () => ({
  name: 'Page',
  render: h => h('section'),
}))

// Mock routes
const mockRoutes = [
  {
    path: ROUTES_PATH.WORKSPACE,
    name: ROUTES_NAMES.WORKSPACE,
    component: WorkspaceList,
    props: true,
  },
]
// Tests

describe('Workspace', () => {
  // For vue router
  const localVue = createLocalVue()
  localVue.use(VueRouter)
  localVue.use(abilitiesPlugin, lckAbilitiesService)

  const router = new VueRouter({ routes: mockRoutes })

  // Default workspace component configuration
  const globalComponentParams = {
    localVue,
    router,
    propsData: { groupId: '0' },
    mocks: {
      t: key => key,
      $t: key => key,
      $toast: {
        add: jest.fn(),
      },
    },
  }

  describe('Create workspace button', () => {
    it('Hide if the user has a USER_PROFILE "USER"', async () => {
      authState.data.user.profile = USER_PROFILE.USER
      const wrapper = await shallowMount(WorkspaceList, globalComponentParams)
      await Vue.nextTick()
      expect(wrapper.findAll('.workspaces-new').length).toBe(0)
      expect(wrapper).toMatchSnapshot()
      authState.data.user.profile = ''
    })
    it('Show if the user is a USER_PROFILE "CREATOR"', async () => {
      authState.data.user.profile = USER_PROFILE.CREATOR
      authState.data.user.rules = [{
        action: 'create',
        subject: 'workspace',
      }]
      lckAbilitiesService.update(authState.data.user.rules)
      const wrapper = await shallowMount(WorkspaceList, globalComponentParams)
      await Vue.nextTick()
      expect(wrapper.findAll('.workspaces-new').length).toBe(1)
      expect(wrapper).toMatchSnapshot()
      authState.data.user.profile = ''
    })
    it('Show if the user is a USER_PROFILE "ADMIN"', async () => {
      authState.data.user.profile = USER_PROFILE.ADMIN
      authState.data.user.rules = [{
        action: 'create',
        subject: 'workspace',
      }]
      lckAbilitiesService.update(authState.data.user.rules)
      const wrapper = await shallowMount(WorkspaceList, globalComponentParams)
      await Vue.nextTick()
      expect(wrapper.findAll('.workspaces-new').length).toBe(1)
      expect(wrapper).toMatchSnapshot()
      authState.data.user.profile = ''
    })
    it('Show if the user is a USER_PROFILE "SUPERADMIN"', async () => {
      authState.data.user.profile = USER_PROFILE.SUPERADMIN
      authState.data.user.rules = [{
        action: 'create',
        subject: 'workspace',
      }]
      lckAbilitiesService.update(authState.data.user.rules)
      const wrapper = await shallowMount(WorkspaceList, globalComponentParams)
      await Vue.nextTick()
      expect(wrapper.findAll('.workspaces-new').length).toBe(1)
      expect(wrapper).toMatchSnapshot()
      authState.data.user.profile = ''
    })
  })
})
