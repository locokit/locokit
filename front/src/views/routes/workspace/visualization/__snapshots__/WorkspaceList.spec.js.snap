// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Workspace Create workspace button Hide if the user has a USER_PROFILE "USER" 1`] = `
<layout-with-header-stub chapters="" groupid="0">
  <div class="generic-view-container p-mx-auto">
    <h1 class="p-my-4 lck-color-title">
      pages.workspace.title
    </h1>
    <div class="p-grid">
      <!---->
      <!---->
      <div class="p-col-12 p-md-4 p-lg-3 workspaces-item">
        <div class="workspaces-button p-mr-2" style="background-color: inherit;">
          <router-link-stub to="/workspace/0" tag="a" ariacurrentvalue="page" event="click" class="workspaces-detail">
            <p class="workspaces-detail-title">workspace-title</p>
          </router-link-stub>
          <!---->
          <!---->
        </div>
      </div>
      <!---->
    </div>
    <p-dialog-stub header="pages.workspace.form.create" modal="true" closable="true" closeonescape="true" showheader="true" basezindex="0" autozindex="true" ariacloselabel="close" position="center">
      <lck-workspace-form-stub></lck-workspace-form-stub>
    </p-dialog-stub>
  </div>
</layout-with-header-stub>
`;

exports[`Workspace Create workspace button Show if the user is a USER_PROFILE "ADMIN" 1`] = `
<layout-with-header-stub chapters="" groupid="0">
  <div class="generic-view-container p-mx-auto">
    <h1 class="p-my-4 lck-color-title">
      pages.workspace.title
    </h1>
    <div class="p-grid">
      <!---->
      <!---->
      <div class="p-col-12 p-md-4 p-lg-3 workspaces-item">
        <div class="workspaces-button p-mr-2" style="background-color: inherit;">
          <router-link-stub to="/workspace/0" tag="a" ariacurrentvalue="page" event="click" class="workspaces-detail">
            <p class="workspaces-detail-title">workspace-title</p>
          </router-link-stub>
          <!---->
          <!---->
        </div>
      </div>
      <div class="p-col-12 p-md-4 p-lg-3 workspaces-item"><button class="workspaces-new"><i class="bi bi-file-plus workspaces-new-icon"></i>
          <p class="p-button p-button-sm">pages.workspace.form.new</p>
        </button></div>
    </div>
    <p-dialog-stub header="pages.workspace.form.create" modal="true" closable="true" closeonescape="true" showheader="true" basezindex="0" autozindex="true" ariacloselabel="close" position="center">
      <lck-workspace-form-stub></lck-workspace-form-stub>
    </p-dialog-stub>
  </div>
</layout-with-header-stub>
`;

exports[`Workspace Create workspace button Show if the user is a USER_PROFILE "CREATOR" 1`] = `
<layout-with-header-stub chapters="" groupid="0">
  <div class="generic-view-container p-mx-auto">
    <h1 class="p-my-4 lck-color-title">
      pages.workspace.title
    </h1>
    <div class="p-grid">
      <!---->
      <!---->
      <div class="p-col-12 p-md-4 p-lg-3 workspaces-item">
        <div class="workspaces-button p-mr-2" style="background-color: inherit;">
          <router-link-stub to="/workspace/0" tag="a" ariacurrentvalue="page" event="click" class="workspaces-detail">
            <p class="workspaces-detail-title">workspace-title</p>
          </router-link-stub>
          <!---->
          <!---->
        </div>
      </div>
      <div class="p-col-12 p-md-4 p-lg-3 workspaces-item"><button class="workspaces-new"><i class="bi bi-file-plus workspaces-new-icon"></i>
          <p class="p-button p-button-sm">pages.workspace.form.new</p>
        </button></div>
    </div>
    <p-dialog-stub header="pages.workspace.form.create" modal="true" closable="true" closeonescape="true" showheader="true" basezindex="0" autozindex="true" ariacloselabel="close" position="center">
      <lck-workspace-form-stub></lck-workspace-form-stub>
    </p-dialog-stub>
  </div>
</layout-with-header-stub>
`;

exports[`Workspace Create workspace button Show if the user is a USER_PROFILE "SUPERADMIN" 1`] = `
<layout-with-header-stub chapters="" groupid="0">
  <div class="generic-view-container p-mx-auto">
    <h1 class="p-my-4 lck-color-title">
      pages.workspace.title
    </h1>
    <div class="p-grid">
      <!---->
      <!---->
      <div class="p-col-12 p-md-4 p-lg-3 workspaces-item">
        <div class="workspaces-button p-mr-2" style="background-color: inherit;">
          <router-link-stub to="/workspace/0" tag="a" ariacurrentvalue="page" event="click" class="workspaces-detail">
            <p class="workspaces-detail-title">workspace-title</p>
          </router-link-stub>
          <!---->
          <!---->
        </div>
      </div>
      <div class="p-col-12 p-md-4 p-lg-3 workspaces-item"><button class="workspaces-new"><i class="bi bi-file-plus workspaces-new-icon"></i>
          <p class="p-button p-button-sm">pages.workspace.form.new</p>
        </button></div>
    </div>
    <p-dialog-stub header="pages.workspace.form.create" modal="true" closable="true" closeonescape="true" showheader="true" basezindex="0" autozindex="true" ariacloselabel="close" position="center">
      <lck-workspace-form-stub></lck-workspace-form-stub>
    </p-dialog-stub>
  </div>
</layout-with-header-stub>
`;
